src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java

```diff
@@ -0,0 +1,162 @@
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringTokenizer;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.plugins.scala.lang.psi.api.ScalaPsiElement;
+import org.jetbrains.plugins.scala.lang.psi.api.ScalaRecursiveElementVisitor;
+import org.jetbrains.plugins.scala.lang.psi.api.base.ScPrimaryConstructor;
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScClass;
+import org.jetbrains.plugins.scala.lang.psi.impl.toplevel.templates.ScExtendsBlockImpl;
+import org.jetbrains.plugins.scala.lang.psi.impl.toplevel.templates.ScTemplateParentsImpl;
+
+public class ClassDeclarationListener implements ActivitiesListener, DocumentListener {
+
+  private final ListenerCallback callback;
+  private final Project project;
+  private final String className;
+  private String[] arguments;
+  private String[] hierarchy;
+  private Disposable disposable;
+
+  /**
+   * Constructor.
+   * @param callback The callback for when the task is complete
+   * @param project The project where the Tutorial is happening
+   * @param className The class name
+   * @param arguments The arguments of the class
+   */
+  public ClassDeclarationListener(ListenerCallback callback,
+                                  Project project,
+                                  String className,
+                                  String[] arguments,
+                                  String[] hierarchy,
+                                  String fileName) {
+    this.callback = callback;
+    this.project = project;
+    this.className = className;
+    this.arguments = arguments;
+    this.hierarchy = hierarchy;
+    Path modulePath = Paths.get(project.getBasePath() + fileName);
+    VirtualFile vf = LocalFileSystem.getInstance().findFileByIoFile(modulePath.toFile());
+    checkPsiFile(PsiManager.getInstance(project).findFile(vf));
+  }
+
+  @Override
+  public boolean registerListener() {
+    EditorEventMulticaster multicaster = EditorFactory.getInstance().getEventMulticaster();
+    disposable = Disposer.newDisposable();
+    multicaster.addDocumentListener(this, disposable);
+    //make sure the correct class file is open?
+    return false;
+  }
+
+  @Override
+  public void unregisterListener() {
+    EditorEventMulticaster multicaster = EditorFactory.getInstance().getEventMulticaster();
+    multicaster.removeDocumentListener(this);
+    disposable.dispose();
+  }
+
+  @Override
+  public void documentChanged(@NotNull DocumentEvent event) {
+    PsiDocumentManager.getInstance(project).commitDocument(event.getDocument());
+    PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(event.getDocument());
+    checkPsiFile(psiFile);
+  }
+
+  private void checkPsiFile(PsiFile psiFile) {
+    if (psiFile != null) {
```

Comment:
move

Reply:
please

Reply:
I don't know what @jaakkonarhi means here, but you should annotate `psiFile` as `@Nullable`. And changing this to `if (psiFile == null) return;` makes the method easier to read.